velox/functions/sparksql/ConcatWs.cpp

```diff
@@ -0,0 +1,391 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/functions/sparksql/ConcatWs.h"
+#include "velox/expression/ConstantExpr.h"
+#include "velox/expression/VectorFunction.h"
+
+namespace facebook::velox::functions::sparksql {
+
+namespace {
+class ConcatWs : public exec::VectorFunction {
+ public:
+  explicit ConcatWs(const std::optional<std::string>& separator)
+      : separator_(separator) {}
+
+  bool isConstantSeparator() const {
+    return separator_.has_value();
+  }
+
+  // Calculate the total number of bytes in the result.
+  size_t calculateTotalResultBytes(
+      const SelectivityVector& rows,
+      exec::EvalCtx& context,
+      std::vector<exec::LocalDecodedVector>& decodedArrays,
```

Comment:
If we add `const`, compiler reports error. So let's keep it unchanged. Thanks!
```
cannot convert ‘const value_type*’ {aka ‘const facebook::velox::exec::LocalDecodedVector*’} to ‘facebook::velox::exec::LocalDecodedVector*’
   54 |         auto indices = decodedArrays[i].get()->indices();

```

Reply:
can you try const auto indices = decodedArrays[i].get()->indices();
Maybe because the function `indices()` return value is const. We do not change it in this function.